src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala

```diff
@@ -3,54 +3,113 @@ package cognite.spark.v1
 import cats.effect.IO
 import cats.implicits._
 import cognite.spark.v1.DataModelInstanceRelation._
-import com.cognite.sdk.scala.common.{CdpApiException, Items}
+import com.cognite.sdk.scala.common.{
+  CdpApiException,
+  DSLAndFilter,
+  DSLEqualsFilter,
+  DSLExistsFilter,
+  DSLInFilter,
+  DSLNotFilter,
+  DSLOrFilter,
+  DSLPrefixFilter,
+  DSLRangeFilter,
+  DomainSpecificLanguageFilter,
+  EmptyFilter
+}
 import com.cognite.sdk.scala.v1._
 import fs2.Stream
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.sources._
 import org.apache.spark.sql.types._
 import org.apache.spark.sql.{Row, SQLContext}
-
 import java.time._
 
-class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)(
-    val sqlContext: SQLContext)
+import com.cognite.sdk.scala.v1.DataModelType.NodeType
+
+// scalastyle:off cyclomatic.complexity
+class DataModelInstanceRelation(
+    config: RelationConfig,
+    spaceExternalId: String,
```

Comment:
Hmmm, but I didn't see any request not requiring this. I don't know why it is optional in. the SDK. Any thoughts @silvavelosa ? Why does `DataModelIdentifier` has `space` optional?

 only `delete` request doesn't require model and space externalIds, it only wants the node/edge externalId.

Reply:
Space is optional because you _can_ omit it (it will then use the blank space, i.e the `default` space). In other words, I think we actually need to make it optional, so people can actually be able to read from and write to the default space

Reply:
I didn't know 😱  Actually, for example, both model and space externalId is not necessary for `delete`. Should I make all optional and check in the specific functions?

And  Do I pass empty string if the user does not provide `spaceExternalId` in other cases?

Reply:
I think spaceExtermalId needs to be optional always, yes.. for how to pass the `default` to Scala SDK I refer to @silvavelosa !

Reply:
as far as I understand, there are models without splace (the documentation refers to `Edge` and `Node`... but maybe there could be more)

Reply:
Yes pretty sure it's possible, but how does one pass that to the SDK if spaceExternalId is non-optional? Empty string, `""`, not possible?

Reply:
So there are three parameters here:
`spaceExternalId`: model space  (used in `DataModelIdentifier` and optional)
`modelExternalId`: model externalId (required when writing)
`instanceSpaceExternalId`:  instance space (required when writing)

Is this correct? I think I mixed two spaces in the code.

Reply:
for models, the `DataModelIdentifier` has optional `spaceExternalId`... for instances, I haven't tried to create nodes without a space, according to the documentation it's required

Reply:
No  I didn't mix, I just need to make `spaceExternalId` optional

Reply:
That's correct Emel, to write a Node you need to supply both `instanceSpaceExternalId` and model Identifier (`spaceExternalId` + `modelExternalId`). Looks like you're correct also @silvavelosa, you can't write nodes into the default space :thinking:

Reply:
Everyone's correct apparently.. good team! :clap:

Reply:
yep, correct! 

`spaceExternalId` in `DataModelIdentifier` is optional, because the documentation refers to models without space

`spaceExternalId` when ingesting nodes/edges is required according to the documentation (it refers to the instance space, not the model space)

Reply:
This is so confusing...